src/Mafia/Shell.hs

```diff
@@ -0,0 +1,373 @@
+
+import qualified Data.Attoparsec.Text as Atto
+import qualified Data.Map as Map
+import qualified Data.Text as T
+
+import           Mafia.Hash
+import           Mafia.Home
+import           Mafia.IO
+import           Mafia.Path
+import           Mafia.Process
+
+import           P
+
+import           System.IO (IO)
+
+import           X.Control.Monad.Trans.Either (EitherT, hoistMaybe, hoistEither)
+
+
+data Script =
+  Script {
+      scriptId :: !Text
+    , scriptHash :: !Hash
+    , scriptPath :: !File
+    , scriptPackages :: ![Package]
+    , scriptSubmodules :: ![Submodule]
+    , scriptMain :: !Text
+    } deriving (Eq, Ord, Show)
+
+newtype Package =
+  Package {
+      unPackage :: Text
+    } deriving (Eq, Ord, Show)
+
+data Submodule =
+  Submodule {
+      submoduleUser :: !Text
+    , submoduleRepo :: !Text
+    , submoduleCommit :: !(Maybe Text)
+    } deriving (Eq, Ord, Show)
+
+data Pragma =
+    PragmaPackage !Package
+  | PragmaSubmodule !Submodule
+    deriving (Eq, Ord, Show)
+
+data ShellError =
+    ShellProcessError !ProcessError
+  | ShellFileNotFound !File
+  | ShellGitParseError ![Char]
+    deriving (Show)
+
+renderShellError :: ShellError -> Text
+renderShellError = \case
+  ShellProcessError err ->
+    renderProcessError err
+
+  ShellFileNotFound path ->
+    "File not found: " <> path
+
+  ShellGitParseError err ->
+    "Failed parsing git output: " <> T.pack err
+
+pPragma :: Text -> Atto.Parser a -> Atto.Parser a
+pPragma pragma parser = do
+  _ <- Atto.string "{-#" *> Atto.skipSpace *> Atto.string pragma
+  x <- Atto.skipSpace *> parser <* Atto.skipSpace
+  _ <- Atto.string "#-}"
+  pure x
+
+pPackage :: Atto.Parser Pragma
+pPackage =
+  fmap PragmaPackage . pPragma "PACKAGE" $
+    Package . T.strip <$> Atto.takeWhile (/= '#')
+
+pAlphaNum :: Atto.Parser Text
+pAlphaNum =
+  Atto.takeWhile1 (Atto.inClass "a-zA-Z0-9")
+
+pSubmodule :: Atto.Parser Pragma
+pSubmodule =
+  fmap PragmaSubmodule . pPragma "SUBMODULE" $
+    Submodule
+      <$> pAlphaNum <* Atto.char '/'
+      <*> pAlphaNum
+      <*> optional (Atto.char '@' *> pAlphaNum)
+
+tryParsePragma :: Text -> Maybe Pragma
+tryParsePragma =
+  rightToMaybe . Atto.parseOnly (Atto.try pPackage <|> pSubmodule)
+
+takePackage :: Pragma -> Maybe Package
+takePackage = \case
+  PragmaPackage x ->
+    Just x
+  _ ->
+    Nothing
+
+takeSubmodule :: Pragma -> Maybe Submodule
+takeSubmodule = \case
+  PragmaSubmodule x ->
+    Just x
+  _ ->
+    Nothing
+
+parseScript :: File -> Text -> Script
+parseScript path source =
+  let
+    pragmas =
+      mapMaybe tryParsePragma $
+      T.lines source
+
+    packages =
+      mapMaybe takePackage pragmas
+
+    submodules =
+      mapMaybe takeSubmodule pragmas
+
+    sid =
+      renderHash $ hashText path
+  in
+    Script sid (hashText source) path packages submodules source
+
+renderSubmoduleUrl :: Submodule -> Text
+renderSubmoduleUrl s =
+  "git@github.com:" <> submoduleUser s <> "/" <> submoduleRepo s
```

Comment:
> saw the same feature in `bower` and it seems to work, although it assumes `https` over `git`

I assumed ssh so that it will work nicely with our private repos, not sure though